### DISCUSSION TITLE
How can I use paths an filenames containing variables?

### DISCUSSION THREAD

**Discussion by lwbt** [twpayne/chezmoi#3485] [Q&A]
Hello, I would like to include a file like the following into my dotfiles:

```bash
.local/share/Steam/userdata/${steam_account_id}/config/preferences_${controller_serial_number}.vdf
```

- `${steam_account_id}` is a large number
- `${controller_serial_number}` is an alphanumerical string
  - Serial number in this particular case a is the serial number of the host device (a Steam Deck) and printed at least on the outside of the cardboard box it was shipped in.

After checking the [documentation](https://www.chezmoi.io/user-guide/templating/#debugging-templates) and threads like https://github.com/twpayne/chezmoi/discussions/3333 and https://github.com/twpayne/chezmoi/issues/1226 it is not clear to me how to do it. I'd prefer a simple way like having `steam_account_id` and `controller_serial_number` in my chezmoi data section and use a paths with variables like above.

**Comment by Lockszmith-GH** [comment]
Not sure if this is the only approach, but seems to me you want to preserve the contendt of the `.vdf` file, but have it a different name on different machines / user profiles.

If that's the goal, one approach can be to maintain a `master_controller_preference.vdf` file and create a symbolic-link to it with a shell (or PowerShell) script with a `run_once_...` script.

**Comment by twpayne** [comment]
There's currently no way to do this elegantly. I would recommend using a `run_` script as @Lockszmith-GH suggests. This can be a template:

```
#!/bin/bash
# .local/share/chezmoi/run_configure-steam-controllers.tmpl

mkdir -p "$HOME/.local/share/Steam/{{ .steamAccountID }}"
cp .local/share/chezmoi/.preferences.vdf "$HOME/.local/share/Steam/{{ .steamAccountID }}/preferences_{{ .controllerSerialNumber}.vdf"
```

Where `.steamAccountID` and `.controllerSerialNumber` are template variables set in the `data` section of your config file, and `.local/share/chezmoi/.preferences.vdf` is a file containing the desired preferences.

This should be a `run_` script, not a `run_once_` script, as it needs to run every time to update the files in case they have changed.

Longer term, #2273 will be the eventual elegant solution.

**Comment by lwbt** [comment]
Thank you for your suggestions and examples.

To give a more complete example, let's say I have more than one file:

```
.local/share/Steam/userdata/${steam_account_id}/config/preferences_${controller_serial_number}.vdf
.local/share/Steam/userdata/${steam_account_id}/config/a_few_more_files.vdf
```

Combining both of your answers I thought of a script like this: **run_*after*_configure-steam-controllers.tmpl**

```bash
#!/bin/bash

# Create a symlink from folder with placeholder name to the actual account ID for steam
ln -sf "__steamAccountID" "$HOME/.local/share/Steam/{{ .steamAccountID }}"

# Create symlinks for files with placeholders in their names in the same directory
ln -sf "preferences__controllerSerialNumber.vdf" \
  "$HOME/.local/share/Steam/{{ .steamAccountID }}/preferences_{{ .controllerSerialNumber}.vdf"
```

So I let Chezmoi copy the folder and all its files with a placeholder name on the same level in the folder hierarchy. Then I create a symbolic link with the proper name that the application expects, and then I create symbolic links for individual files where necessary. It's semi-elegant, or just good enough.

I spent some more time catching up with the changes in documentation and some discussions here. I read that symbolic links for folders are not read in the source state for technical reasons and security, but on the targets this should be fine? Or will I run into some other issues of Chezmoi?

**Reply by Lockszmith-GH** [reply]
You can do both copy and symlink, or just copy actually.

With [twpayne's suggestion](https://github.com/twpayne/chezmoi/discussions/3485#discussioncomment-8146750) you don't need symlinks at all actually, all you need is a directory `.preferences.vdf` in your chezmoi git repo, it will not be copied over automatically.

With the symlink approach, you are right - a copy will be managed by chezmoi somewhere in your home directory, and the symlinks will point to files/folders.

Note, in your code, after you symlink the directory, it will fill that directory with additioal (symlinked) 'files' - not sure that is a desired effect. So that makes tpayne's solution a bit 'more elegant'.

Since these are generally small files, I think you should be safe with the small files. But keep in mind that you should always edit the git repo copy, and then `chezmoi apply` to apply the changes.
